src/transport/SessionManager.cpp

```diff
@@ -773,29 +786,59 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
         return;
     }
 
-    System::PacketBufferHandle msgCopy;
+    // Extract MIC from the end of the message.
+    uint8_t * data     = msg->Start();
+    uint16_t len       = msg->DataLength();
+    uint16_t footerLen = packetHeader.MICTagLength();
+    VerifyOrReturn(footerLen <= len);
+
+    uint16_t taglen = 0;
+    MessageAuthenticationCode mac;
+    ReturnOnFailure(mac.Decode(packetHeader, &data[len - footerLen], footerLen, &taglen));
+    VerifyOrReturn(taglen == footerLen);
+
     bool decrypted = false;
     while (!decrypted && iter->Next(groupContext))
     {
+        CryptoContext context(groupContext.keyContext);
+        msgCopy = msg.CloneData();
+
+        if (packetHeader.HasPrivacyFlag())
+        {
+            // Perform privacy deobfuscation, if applicable.
+            uint8_t * privacyHeader = msgCopy->Start() + Header::kPrivacyHeaderOffset;
+            size_t privacyLength    = packetHeader.PrivacyHeaderLength();
+            ReturnOnFailure(context.PrivacyDecrypt(privacyHeader, privacyLength, privacyHeader, packetHeader, mac));
+        }
+
+        ReturnOnFailure(packetHeaderCopy.DecodeAndConsume(msgCopy));
+
+        if (msgCopy.IsNull())
```

Comment:
This check doesn't make sense here.  If `msgCopy` were null, we would have crashed before this point.

Reply:
That does not make sense to me.  `DecodeAndConsume` is not going to make the handle null....

Reply:
Removed as advised.